rgw/rgw.go

```diff
@@ -0,0 +1,805 @@
+		C.uint32_t(posixFlags), C.uint32_t(flags))
+	if ret == 0 {
+		st := syscall.Stat_t{
+			Dev:     uint64(stat.st_dev),
+			Ino:     uint64(stat.st_ino),
+			Nlink:   uint64(stat.st_nlink),
+			Mode:    uint32(stat.st_mode),
+			Uid:     uint32(stat.st_uid),
+			Gid:     uint32(stat.st_gid),
+			Rdev:    uint64(stat.st_rdev),
+			Size:    int64(stat.st_size),
+			Blksize: int64(stat.st_blksize),
+			Blocks:  int64(stat.st_blocks),
+			Atim: syscall.Timespec{
+				Sec:  int64(stat.st_atim.tv_sec),
+				Nsec: int64(stat.st_atim.tv_nsec),
+			},
+			Mtim: syscall.Timespec{
+				Sec:  int64(stat.st_mtim.tv_sec),
+				Nsec: int64(stat.st_mtim.tv_nsec),
+			},
+			Ctim: syscall.Timespec{
+				Sec:  int64(stat.st_ctim.tv_sec),
+				Nsec: int64(stat.st_ctim.tv_nsec),
+			},
+		}
+		return fh, &st, nil
+	}
+	return nil, nil, getError(ret)
+}
+
+// MkdirFlag is used to control behavor of Mkdir()
+type MkdirFlag uint32
+
+const (
+	// MkdirFlagNone keep the default behavior of Mkdir()
+	MkdirFlagNone MkdirFlag = 0
+)
+
+// Mkdir creates a new directory
+//
+// Implements:
+//    int rgw_mkdir(rgw_fs *fs,
+//                  rgw_file_handle *parent_fh,
+//                  const char *name, stat *st, uint32_t mask,
+//                  rgw_file_handle **fh, uint32_t flags)
+//
+func (fs *FS) Mkdir(parentHdl *FileHandle, name string, mask AttrMask, flags MkdirFlag) (
+	*FileHandle, *syscall.Stat_t, error) {
+
+	fh := &FileHandle{}
+	var stat C.struct_stat
+
+	cName := C.CString(name)
+	defer C.free(unsafe.Pointer(cName))
+
+	ret := C.rgw_mkdir(fs.rgwFS, parentHdl.handle, cName, &stat, C.uint32_t(mask), &fh.handle,
+		C.uint32_t(flags))
+	if ret == 0 {
+		st := syscall.Stat_t{
+			Dev:     uint64(stat.st_dev),
+			Ino:     uint64(stat.st_ino),
+			Nlink:   uint64(stat.st_nlink),
+			Mode:    uint32(stat.st_mode),
+			Uid:     uint32(stat.st_uid),
+			Gid:     uint32(stat.st_gid),
+			Rdev:    uint64(stat.st_rdev),
+			Size:    int64(stat.st_size),
+			Blksize: int64(stat.st_blksize),
+			Blocks:  int64(stat.st_blocks),
+			Atim: syscall.Timespec{
+				Sec:  int64(stat.st_atim.tv_sec),
+				Nsec: int64(stat.st_atim.tv_nsec),
+			},
+			Mtim: syscall.Timespec{
+				Sec:  int64(stat.st_mtim.tv_sec),
+				Nsec: int64(stat.st_mtim.tv_nsec),
+			},
+			Ctim: syscall.Timespec{
+				Sec:  int64(stat.st_ctim.tv_sec),
+				Nsec: int64(stat.st_ctim.tv_nsec),
+			},
+		}
+		return fh, &st, nil
+	}
+	return nil, nil, getError(ret)
+}
+
+// WriteFlag is used to control behavor of Write()
+type WriteFlag uint32
+
+const (
+	// WriteFlagNone keep the default behavior of Write()
+	WriteFlagNone WriteFlag = 0
+)
+
+// Write data to file
+//
+// Implements:
+//    int rgw_write(rgw_fs *fs,
+//                  rgw_file_handle *fh, uint64_t offset,
+//                  size_t length, size_t *bytes_written, void *buffer,
+//                  uint32_t flags)
+//
+func (fs *FS) Write(fh *FileHandle, buffer []byte, offset uint64, length uint64,
```

Comment:
Methods with common names to Go interfaces should probably match the common interfaces. For example io.Writer defines Write as `Write(p []byte) (n int, err error)`. On top of that, we want the various sub-packages of go-ceph to be reasonably consistent. Therefore I think that this would be better named as WriteAt and exist as a method on the FileHandle type.